include/rocksdb/data_structure.h

```diff
@@ -45,4 +47,114 @@ class SmallEnumSet {
  private:
   uint64_t state_;
 };
+
+// An abstract class for a sequence of items that are materialized
+// on demand. This allows for a convenient, memory-inefficient public
+// interface for each item while the full sequence data can be stored
+// in a memory-efficient way with hidden details. For simplicity,
+// foreign language bindings can treat this structure as a vector by
+// immediately materializing all the elements with ToVector().
+template <typename Item>
```

Comment:
I am not sure I understand why we want this class, versus just using a std::vector.  It seems like it ends up in the public API in a way that might create more confusion for users.

Reply:
Lets say the directory string is 50 bytes and our DB has 10,000 SST files. In the common case of all files in the same directory, is it reasonable to hold redundant 500KB of string data? Add more to that if there are checksums. Perhaps more data added in future. These objects will almost certainly stay alive while making a checkpoint or backup, etc. which can be a slow operation. What is the threshold for memory worth tracking?

Obviously my inclination is to reserve the option to keep the data for all files normalized, for memory efficiency, while denormalizing only on demand for convenience.